File: src/syntax/parser/token_sink.cc
#include "syntax/parser/token_sink.h"

#include <ranges>
#include <stdexcept>
#include <variant>
#include <vector>

#include "syntax/parser/event.h"
#include "syntax/syntax_kind.h"

namespace orion::syntax {
TokenSink::Result TokenSink::Finish() {
  for (size_t event_idx = 0; event_idx < events_.size(); event_idx++) {
    if (const Event event = events_[event_idx];
        event.HoldsAlternative<Event::Start>()) {
      const auto e = event.Get<Event::Start>();
      events_[event_idx] = Event::CreatePlaceholder();

      StartNode(event_idx, e.Kind(), e.ForwardParent());
    } else if (event.HoldsAlternative<Event::Finish>()) {
      FinishNode();
    } else if (event.HoldsAlternative<Event::Token>()) {
      AddToken();
    } else if (event.HoldsAlternative<Event::Placeholder>()) {
      // Do Nothing
    } else {
      throw std::invalid_argument("unknown event type");
    }
  }

  return TokenSink::Result{builder_.Finish()};
}

void TokenSink::StartNode(const size_t event_idx, const SyntaxKind kind,
                          const std::optional<size_t> forward_parent) {
  size_t event_idx_mut = event_idx;
  std::optional<size_t> forward_parent_mut = forward_parent;

  std::vector kinds = {kind};
  while (forward_parent_mut.has_value()) {
    event_idx_mut += forward_parent_mut.value();

    if (const Event event = events_[event_idx_mut];
        event.HoldsAlternative<Event::Start>()) {
      const auto e = event.Get<Event::Start>();
      events_[event_idx_mut] = Event::CreatePlaceholder();

      kinds.emplace_back(e.Kind());
      forward_parent_mut = e.ForwardParent();
    } else {
      throw std::invalid_argument("unreachable event in StartNode");
    }
  }

  for (auto& k : std::ranges::reverse_view(kinds)) {
    builder_.StartNode(k);
  }
}
}  // namespace orion::syntax

File: crates/yuzu_parser/src/token_sink.rs
use std::mem;

use rowan::{GreenNode, GreenNodeBuilder, Language};
use yuzu_diagnostics::diagnostics::{Diagnostic, engine::DiagnosticsEngine};
use yuzu_lexer::{lexer::Token, token_kind::TokenKind};
use yuzu_syntax::{SyntaxKind, YuzuLanguage};

use crate::parser::event::Event;

pub(crate) struct Result {
    pub(crate) green: GreenNode,
}

pub(crate) struct TokenSink<'t, 'input> {
    builder: GreenNodeBuilder<'t>,
    tokens: &'t [Token<'input>],
    cursor: usize,
    events: Vec<Event>,
    diagnostics: &'t mut DiagnosticsEngine,
}

impl<'t, 'input> TokenSink<'t, 'input> {
    pub(crate) fn new(
        tokens: &'t [Token<'input>],
        events: Vec<Event>,
        diagnostics: &'t mut DiagnosticsEngine,
    ) -> Self {
        Self {
            builder: GreenNodeBuilder::default(),
            tokens,
            cursor: 0,
            events,
            diagnostics,
        }
    }

    pub(crate) fn finish(mut self) -> Result {
        let mut depth = 0usize;
        for idx in 0..self.events.len() {
            match mem::replace(&mut self.events[idx], Event::Placeholder) {
                Event::Start {
                    kind,
                    forward_parent,
                } => {
                    if depth > 0 {
                        self.bump_trivia();
                    }
                    depth += self.start(idx, kind, forward_parent);
                }
                Event::Token => {
                    self.bump_trivia();
                    self.token();
                }
                Event::Finish => {
                    if depth == 1 {
                        self.bump_trivia();
                    }
                    depth -= 1;
                    self.builder.finish_node();
                }
                Event::Error { error } => {
                    let diagnostic: Diagnostic = error.into();
                    self.diagnostics.emit(diagnostic);
                }
                Event::Placeholder => {}
            }
        }

        Result {
            green: self.builder.finish(),
        }
    }

    fn start(&mut self, idx: usize, kind: SyntaxKind, forward_parent: Option<usize>) -> usize {
        let mut kinds = vec![kind];

        let mut idx = idx;
        let mut forward_parent = forward_parent;

        // Walk through the forward parent of the forward parent and the forward parent
        // of that, and of that, etc. until we reach a StartNode event without a forward
        // parent.
        while let Some(fp) = forward_parent {
            idx += fp;

            forward_parent = if let Event::Start {
                kind,
                forward_parent,
            } = mem::replace(&mut self.events[idx], Event::Placeholder)
            {
                kinds.push(kind);
                forward_parent
            } else {
                unreachable!()
            };
        }

        let opened = kinds.len();
        for kind in kinds.into_iter().rev() {
            self.builder.start_node(YuzuLanguage::kind_to_raw(kind));
        }
        opened
    }

    fn token(&mut self) {
        let Token { kind, text, .. } = self.tokens[self.cursor];
        self.builder
            .token(YuzuLanguage::kind_to_raw(kind.into()), text);
        self.cursor += 1;
    }

    fn bump_trivia(&mut self) {
        while self.cursor < self.tokens.len() {
            let token = &self.tokens[self.cursor];
            if !TokenKind::is_trivia(token.kind) {
                break;
            }

            self.token();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use text_size::TextRange;
    use yuzu_diagnostics::source_map::SourceMap;
    use yuzu_lexer::lexer::Lexer;
    use yuzu_syntax::SyntaxNode;

    use crate::parser::error::ParseError;

    #[test]
    fn builds_a_tree_from_events() {
        let tokens: Vec<_> = Lexer::new("+").collect();
        let events = vec![
            Event::Start {
                kind: SyntaxKind::Root,
                forward_parent: None,
            },
            Event::Token,
            Event::Finish,
        ];
        let mut diagnostics = DiagnosticsEngine::new();

        let result = TokenSink::new(&tokens, events, &mut diagnostics).finish();
        let root = SyntaxNode::new_root(result.green);

        assert_eq!(root.kind(), SyntaxKind::Root);
        assert_eq!(root.text().to_string(), "+");
        assert!(diagnostics.diagnostics().is_empty());
    }

    #[test]
    fn error_events_are_emitted_as_diagnostics() {
        let tokens: Vec<_> = Lexer::new("+").collect();
        let source_id = {
            let mut sources = SourceMap::new();
            sources.add(String::new(), String::new())
        };
        let error = ParseError::ExpectedExpression {
            found: None,
            range: TextRange::default(),
            source_id,
        };
        let events = vec![
            Event::Start {
                kind: SyntaxKind::Root,
                forward_parent: None,
            },
            Event::Error { error },
            Event::Finish,
        ];
        let mut diagnostics = DiagnosticsEngine::new();

        let _ = TokenSink::new(&tokens, events, &mut diagnostics).finish();

        assert_eq!(diagnostics.diagnostics().len(), 1);
    }
}
